### PR TITLE
Add FasterTokenizer on PPMiniLM

### PR DESCRIPTION
**Pull request by LiuChiachi**
### PR types

 New features
### PR changes

Models | APIs
### Description

Add FasterTokenizer on PPMiniLM

TODO：

- [x] 更新PPMiniLM类，支持训练、导出差异化：
    - [x] 动态图Finetuning、裁剪训练不接入FasterTokenizer正常训练；
    - [x] Finetuning、裁剪的导出支持导出带有FasterTokenizer op的图
- [x] 支持离线量化带有FasterTokenizer op的图

PS：
此PR正常运行需要依赖 https://github.com/PaddlePaddle/Paddle/pull/38686 和https://github.com/PaddlePaddle/PaddleSlim/pull/964  的合入

<!DOCTYPE html><figure class="md-table-fig" cid="n1112" mdtype="table" style="box-sizing: border-box; margin: 1.2em 0px; overflow-x: auto; max-width: calc(100% + 16px); padding: 0px; cursor: default; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;">

Model | #Params | #FLOPs | Speedup | AFQMC | TNEWS | IFLYTEK | CMNLI | OCNLI | CLUEWSC2020 | CSL | CLUE 平均值
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
BERTbase | 102.3M | 10.87B | 1.00x | 74.14 | 56.81 | 61.10 | 81.19 | 74.85 | 79.93 | 81.47 | 72.78
TinyBERT6 | 59.7M | 5.44B | 2.04x | 72.59 | 55.70 | 57.64 | 79.57 | 73.97 | 76.32 | 80.00 | 70.83
UER-py RoBERTa L6-H768 | 59.7M | 5.44B | 2.04x | 69.62 | 66.45 | 59.91 | 76.89 | 71.36 | 71.05 | 82.87 | 71.16
RBT6, Chinese | 59.7M | 5.44B | 2.04x | 73.93 | 56.63 | 59.79 | 79.28 | 73.12 | 77.30 | 80.80 | 71.55
ERNIE-Tiny | 90.7M | 4.83B | 2.30x | 71.55 | 58.34 | 61.41 | 76.81 | 71.46 | 72.04 | 79.13 | 70.11
PP-MiniLM 6L-768H | 59.7M | 5.44B | 2.12x | 74.14 | 57.43 | 61.75 | 81.01 | 76.17 | 86.18 | 79.17 | 73.69
PP-MiniLM 裁剪后 | 49.1M | 4.08B | 2.60x | 73.91 | 57.44 | 61.64 | 81.10 | 75.59 | 85.86 | 78.53 | 73.44
PP-MiniLM 裁剪 + 量化后 | 49.2M | - | 9.26x | 74.00 | 57.37 | 61.33 | 81.09 | 75.56 | 85.85 | 78.57 | 73.40

</figure>

### CODE REVIEW

File: examples/model_compression/pp-minilm/README.md [hunk 1]
```diff
@@ -43,13 +43,13 @@ PP-MiniLM 压缩方案以面向预训练模型的任务无关知识蒸馏(Task-a
...
-| UER-py RoBERTa L6- H768 | 59.7M   | 5.44B  | 1.88x   | 69.62 | 66.45 | 59.91   | 76.89 | 71.36 | 71.05       | 82.87 | 71.16       |
-| RBT6, Chinese           | 59.7M   | 5.44B  | 1.88x   | 73.93 | 56.63 | 59.79   | 79.28 | 73.12 | 77.30       | 80.80 | 71.55       |
-| ERNIE-Tiny              | 90.7M   | 4.83B  | 2.22x   | 71.55 | 58.34 | 61.41   | 76.81 | 71.46 | 72.04       | 79.13 | 70.11       |
-| PP-MiniLM 6L-768H       | 59.7M   | 5.44B  | 1.88x   | 74.14 | 57.43 | 61.75   | 81.01 | 76.17 | 86.18       | 79.17 | 73.69       |
-| PP-MiniLM 裁剪后        | 49.1M   | 4.08B  | 2.39x   | 73.91 | 57.44 | 61.64   | 81.10 | 75.59 | 85.86       | 78.53 | 73.44       |
-| PP-MiniLM 量化后        | 49.2M   | -      | 5.35x   | 74.00 | 57.37 | 61.33   | 81.09 | 75.56 | 85.85       | 78.57 | 73.40       |
+| TinyBERT<sub>6</sub>    | 59.7M   | 5.44B  | 2.04x   | 72.59 | 55.70 | 57.64   | 79.57 | 73.97 | 76.32       | 80.00 | 70.83       |
+| UER-py RoBERTa L6- H768 | 59.7M   | 5.44B  | 2.04x   | 69.62 | 66.45 | 59.91   | 76.89 | 71.36 | 71.05       | 82.87 | 71.16       |
```
Comment (ZeyuChen) [inline examples/model_compression/pp-minilm/README.md]: L6-H768 多了个空格

File: examples/model_compression/pp-minilm/finetuning/run_clue.py [hunk 2]
```diff
@@ -404,6 +409,7 @@ def print_arguments(args):
     print_arguments(args)
     if args.do_train:
         do_train(args)
-        export_model(args)
+        if args.save_inference_model:
```
Comment (tianxin1860) [inline examples/model_compression/pp-minilm/finetuning/run_clue.py]: 为什么需要 save_inference_model 和 save_inference_model_with_tokenizer 2 个命令行参数？

**Review by tianxin1860 — COMMENTED** [review]
需要把 FasterTokenizer 当做 1 个开关给用户控制么？在 FasterTokenizer 比普通 Tokenizer 显著快又没有其它劣势的条件下，我们对外就直接提供最优版本 FasterTokenizer 即可？

File: paddlenlp/transformers/ppminilm/modeling.py [hunk 3]
```diff
@@ -141,6 +153,53 @@ def init_weights(self, layer):
...
+            is_split_into_words=self.ppminilm.is_split_into_words)
+
+    def to_static(self,
+                  output_path,
+                  use_faster_tokenizer=True,
+                  is_text_pair=False,
+                  pad_to_max_seq_len=False):
+        self.eval()
+        self.use_faster_tokenizer = use_faster_tokenizer
+        self.pad_to_max_seq_len = pad_to_max_seq_len
```
Comment (tianxin1860) [inline paddlenlp/transformers/ppminilm/modeling.py]: 传入的 pad_to_max_seq_len 没有使用？

File: paddlenlp/transformers/ppminilm/modeling.py [hunk 4]
```diff
@@ -141,6 +153,53 @@ def init_weights(self, layer):
...
+                    self,
+                    input_spec=[
+                        paddle.static.InputSpec(
+                            shape=[None], dtype=core.VarDesc.VarType.STRINGS),
+                        paddle.static.InputSpec(
+                            shape=[None], dtype=core.VarDesc.VarType.STRINGS)
+                    ])
+            else:
+
+                model = paddle.jit.to_static(
```
Comment (tianxin1860) [inline paddlenlp/transformers/ppminilm/modeling.py:184]: 基类 FasterPretrainedModel 已经实现了 to_static 函数，此处是否直接调用基类函数即可？

File: paddlenlp/transformers/ppminilm/modeling.py [hunk 5]
```diff
@@ -141,6 +153,53 @@ def init_weights(self, layer):
...
+        self.eval()
+        self.use_faster_tokenizer = use_faster_tokenizer
+        self.pad_to_max_seq_len = pad_to_max_seq_len
+        # Convert to static graph with specific input description
+        if self.use_faster_tokenizer:
+            self.add_faster_tokenizer_op()
+            if is_text_pair:
+                model = paddle.jit.to_static(
+                    self,
+                    input_spec=[
```
Comment (tianxin1860) [inline paddlenlp/transformers/ppminilm/modeling.py:174]: 为什么 faster 版本的输入 shape 是 [None], 非 Faster 版本的 shape 是 [None, None]？

Comment (tianxin1860) [reply paddlenlp/transformers/ppminilm/modeling.py:174]: 这里的 to_static 逻辑是放在 FasterPretrainedModel 里还是暴露给 FasterTokenizer 的用户比较合适？@Steffy-zxf
@wawltor @LiuChiachi

File: paddlenlp/transformers/ppminilm/modeling.py [hunk 6]
```diff
@@ -116,14 +120,22 @@ class PPMiniLMPretrainedModel(PretrainedModel):
...
             "ppminilm-6l-768h":
             "https://bj.bcebos.com/paddlenlp/models/transformers/ppminilm-6l-768h/ppminilm-6l-768h.pdparams",
+        },
+        "vocab_file": {
+            "ppminilm-6l-768h":
+            "https://bj.bcebos.com/paddlenlp/models/transformers/ppminilm-6l-768h/vocab.txt",
         }
     }
     base_model_prefix = "ppminilm"
+    use_faster_tokenizer = False
```
Comment (tianxin1860) [inline paddlenlp/transformers/ppminilm/modeling.py:138]: 这个变量加在这里感觉没有起作用？

File: paddlenlp/transformers/ppminilm/modeling.py [hunk 7]
```diff
@@ -304,6 +379,13 @@ def forward(self,
                 sequence_output, pooled_output = model(**inputs)
 
         """
+        # Only for saving
+        if self.use_faster_tokenizer:
+            input_ids, token_type_ids = self.tokenizer(
+                text=input_ids,
+                text_pair=token_type_ids,
+                max_seq_len=self.max_seq_len,
+                pad_to_max_seq_len=self.pad_to_max_seq_len)
```
Comment (tianxin1860) [inline paddlenlp/transformers/ppminilm/modeling.py:402]: 这里的 tokenizer 是指 FasterTokenizer 吧？输入为什么是 input_ids 和 token_type_ids？

**Review by tianxin1860 — COMMENTED** [review]
Leave some comments

File: examples/model_compression/pp-minilm/data.py [hunk 8]
```diff
@@ -34,6 +34,49 @@
 }
 
 
+def get_example_for_faster_tokenizer(example,
```
Comment (tianxin1860) [inline examples/model_compression/pp-minilm/data.py]: 和 convert_exmaple 函数的唯一区别是少了 tokenize 函数么？

**Review by tianxin1860 — COMMENTED** [review]
Leave a comment

File: examples/model_compression/pp-minilm/data.py (hunk 8 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/data.py]: 是的，区别不是很大，我把它们合成一个函数

File: examples/model_compression/pp-minilm/finetuning/run_clue.py (hunk 2 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/finetuning/run_clue.py]: 已经把`save_inference_model_with_tokenizer`去掉了，之前的`use_faster_tokenizer`可以发挥这样的功能

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 6 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py:138]: 默认是False，这样在训练的时候可以走不带`use_faster_tokenizer`的逻辑

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 3 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py]: 动转静forward中有被用到

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 5 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py:174]: `dtype=core.VarDesc.VarType.STRINGS`时shape参数无效

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 4 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py:184]: 加入了`use_faster_tokenizer`以及`pad_to_max_seq_len`2个参数，以及增加了对text pair 为输入的模型的导出，求 @ZeyuChen @guoshengCS 能帮忙看一下～

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 7 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py:402]: 共用了一样的参数列表，`use_faster_tokenizer == True` 时计算真正的`input_ids` 和 `token_type_ids`

Comment (tianxin1860) [reply paddlenlp/transformers/ppminilm/modeling.py:402]: `input_ids` 和 `token_type_ids` 分别表示的含义是 text、text_pair,  为什么不使用语义一致的变量命名？

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 4 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py:184]: 已经删除了`pad_to_max_len`参数，因为不是必要的

File: examples/model_compression/pp-minilm/data.py (hunk 8 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/data.py]: 已经修改，感谢指出：）

File: paddlenlp/transformers/ppminilm/modeling.py (hunk 7 above)
Comment (LiuChiachi) [reply paddlenlp/transformers/ppminilm/modeling.py:402]: 已经修改了docstring

File: examples/model_compression/pp-minilm/README.md (hunk 1 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/README.md]: 感谢指出，已经修改

File: examples/model_compression/pp-minilm/pruning/export_model.py [hunk 9]
```diff
@@ -117,20 +136,10 @@ def parse_args():
...
-        paddle.static.InputSpec(
-            shape=[None, max_seq_length], dtype='int64'),
-        paddle.static.InputSpec(
-            shape=[None, max_seq_length], dtype='int64')
-    ]
-    net = paddle.jit.to_static(model, input_spec=input_shape)
-    paddle.jit.save(net, model_path)
-
-
 def do_train(args):
```
Comment (tianxin1860) [inline examples/model_compression/pp-minilm/pruning/export_model.py]: 函数命名 do_train -> export_model ?

File: examples/model_compression/pp-minilm/pruning/export_model.py [hunk 10]
```diff
@@ -183,11 +193,34 @@ def do_train(args):
...
+                ],
+                origin_model=origin_model)
+        else:
+            origin_model_new = ofa_model.export(
+                best_config,
+                input_shapes=[1],
+                input_dtypes=core.VarDesc.VarType.STRINGS,
+                origin_model=origin_model)
+    else:
+        ofa_model.model.use_faster_tokenizer = args.use_faster_tokenizer
```
Comment (tianxin1860) [inline examples/model_compression/pp-minilm/pruning/export_model.py]: 1. 这里为什么需要 ofa_model.model.use_faster_tokenizer = args.use_faster_tokenizer ? 
2. ofa_model.model.use_faster_tokenizer 默认值我记得设置就是 False 为了避免训练进入 FasterTokenizer。

**Review by tianxin1860 — COMMENTED** [review]
Leave some comments

File: examples/model_compression/pp-minilm/quantization/quant_post.py [hunk 11]
```diff
@@ -106,6 +120,27 @@ def batch_generator_func():
...
+    def batch_generator_func_using_faster_tokenizer():
+        if 'sentence' in dev_ds[0]:
+            batch_data = []
+        else:
+            batch_data = [[], []]
+        for data in dev_ds:
+            if 'sentence' in data:
+                batch_data.append(data['sentence'])
+                if len(batch_data) == batch_size:
+                    yield {"input_ids": batch_data}
```
Comment (tianxin1860) [inline examples/model_compression/pp-minilm/quantization/quant_post.py:132]: 这里传入的是数据 text,  变量名必须固定为 "input_ids" 么？

File: examples/model_compression/pp-minilm/quantization/quant_post.py [hunk 12]
```diff
@@ -106,6 +120,27 @@ def batch_generator_func():
...
+                    batch_data = []
+            else:
+                batch_data[0].append(data['sentence1'])
+                batch_data[1].append(data['sentence2'])
+                if len(batch_data[0]) == batch_size:
+                    yield {
+                        "input_ids": batch_data[0],
+                        "token_type_ids": batch_data[1]
```
Comment (tianxin1860) [inline examples/model_compression/pp-minilm/quantization/quant_post.py]: `input_ids` 和 `token_type_ids` 对应的数据都是 text, 必须固定用这 2 个变量名么？

**Review by tianxin1860 — COMMENTED** [review]
整体看下来，Fine-tune、裁剪、量化部分为了同时支持 FasterTokenizer 版本和普通 Tokenizer 版本，逻辑略显复杂，会引入语义不一致问题，个人建议可以只支持  FasterTokenizer 版本即可，无需支持普通 Tokenizer。

File: examples/model_compression/pp-minilm/quantization/quant_post.py (hunk 11 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/quantization/quant_post.py:132]: 感谢指出，已经修改了导出text时输入的name

File: examples/model_compression/pp-minilm/quantization/quant_post.py (hunk 12 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/quantization/quant_post.py]: 感谢指出，已经修改了导出text时输入的name，改成了"text"和“text_pair“

File: examples/model_compression/pp-minilm/pruning/export_model.py (hunk 9 above)
Comment (LiuChiachi) [reply examples/model_compression/pp-minilm/pruning/export_model.py]: 感谢指出，export_model和脚本名重复，为了维持一样的格式，目前改成了`do_export`

**Review by tianxin1860 — APPROVED** [review]
LGTM
